crypto/bn/bn_prime.c

```diff
@@ -467,3 +467,406 @@ static int probable_prime_dh_safe(BIGNUM *p, int bits, const BIGNUM *padd,
+
+    BN_CTX_start(ctx);
+
+    Q = BN_CTX_get(ctx);
+    D = BN_CTX_get(ctx);
+    U = BN_CTX_get(ctx);
+    V = BN_CTX_get(ctx);
+    p1 = BN_CTX_get(ctx);
+    i = BN_CTX_get(ctx);
+    d = BN_CTX_get(ctx);
+    n1 = BN_CTX_get(ctx);
+    n2 = BN_CTX_get(ctx);
+    if (n2 == NULL)
+        goto done;
+
+    if (!BN_set_word(D, 5) || !BN_one(n1) || !BN_set_word(n2, 2))
+        goto done;
+
+    if (BN_cmp(p, n2) == 0) {
+        /*
+         * 2 is a lucas prime, but Jacobi does not work for odd numbers so
+         * would return an error.
+         */
+        ret = 1;
+        goto done;
+    }
+
+    if (BN_cmp(p, n2) > 0 && !BN_is_odd(p)) {
+        ret = 0;
+        goto done;
+    }
+
+    while (1) {
+        int jacobi = BN_jacobi_symbol(D, p, ctx);
+        if (jacobi == -2)
+            goto done;
+        if (jacobi == -1)
+            break;
+        BN_set_negative(D, 0);
+        /*
+         * 5 and 11 end up with testing themself, and Jacobi then returns 0.
+         * Just continue searching for D instead of saying, it should still find
+         * either a -1 or 0.
+         */
+        if (jacobi == 0 && BN_cmp(D, p) != 0) {
+            ret = 0;
+            goto done;
+        }
+
+        /* On average after 1.78 tries we should have found D, but we won't find
+         * it in case p is a perfect square. So check that it's a perfect square
+         * after the 7th D we tried. */
+        if (BN_is_word(D, 17)) {
+            int ps = BN_is_perfect_square(p, ctx);
+            if (ps == -1)
+                goto done;
+            if (ps == 1) {
+                ret = 0;
+                goto done;
+            }
+        }
+
+        /* D should have the Selfright sequence 5, -7, 9, -11, 13, -15, ... */
+        if (!BN_add(D, D, n2))
+            goto done;
+        neg = !neg;
+        BN_set_negative(D, neg);
+    }
+
+    /* Q = (1-D)/4, so -1, 2, -2, 3, -3, ... */
+    if (BN_copy(Q, D) == NULL)
+        goto done;
+    if (!BN_sub(Q, n1, D))
+        goto done;
+    if (!BN_rshift(Q, Q, 2))
+        goto done;
+
+    if (!BN_add(p1, p, n1))
+            goto done;
+
+    if (BN_copy(d, p1) == NULL)
+        goto done;
+    while(!BN_is_odd(d))
+        if (!BN_rshift1(d, d))
+            goto done;
+
+    if (!BN_one(U) || !BN_one(V) || !BN_one(i))
+        goto done;
+
+    if (!bn_lucas_sequence(Q, D, p, U, V, i, d, ctx) ||
+        !BN_mod(U, U, p, ctx))
+        goto done;
+    if (BN_is_zero(U))
+        strong = 1;
+    if (!BN_mod(V, V, p, ctx))
+        goto done;
+    if (BN_is_zero(V))
+        strong = 1;
+
+    while (BN_ucmp(i, p1) != 0) {
+        if (!BN_lshift1(d, i) ||
+            !bn_lucas_sequence(Q, D, p, U, V, i, d, ctx) ||
+            !BN_mod(V, V, p, ctx))
+            goto done;
+        if (BN_is_zero(V))
+            strong = 1;
+    }
+
+    if (!strong) {
+        ret = 0;
+        goto done;
+    }
+
+    if (!BN_is_zero(U)) {
+        ret = 0;
+        goto done;
+    }
+
+#if 0
```

Comment:
I'm wondering why this is left here?
The checks are very cheap but if they aren't used...

Any idea as to how often these eliminate other non-primes?
Would it be worthwhile having a second function that does the strong Lucas prime plus these extras?  Or perhaps an extra function argument?

Reply:
For that list of strong Lucas primes, for 100K, it leaves 3 I think of the 12 if memory is good.

The reason I keep this around is that at some point I want to move this check to a new prime test function.